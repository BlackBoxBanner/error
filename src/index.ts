/**
 * Represents the structure of a custom error.
 */
export type ErrorType = {
  id?: string;
  message: string;
};

/**
 * Represents a function that constructs a custom error.
 */
type ErrorConstruct = (props: ErrorType) => Error;

/**
 * Represents a function that parses a custom error.
 */
type ErrorParser = (error: Error) => ErrorType;

/**
 * Constructs a custom error with optional properties.
 * @param {ErrorType} { id, message } - The properties of the error.
 * @returns {Error} - The constructed custom error.
 */
export const customError: ErrorConstruct = ({ id = "internal", message }) => {
  return new Error(JSON.stringify({ id, message }));
};

/**
 * Parses a custom error and returns its properties.
 * @param {Error} error - The custom error to parse.
 * @returns {ErrorType} - The parsed error properties.
 * @throws {ErrorType} - Throws a custom error if parsing fails.
 */
export const errorParser: ErrorParser = (error: Error) => {
  try {
    return JSON.parse(error.message) as ErrorType;
  } catch (err: unknown) {
    if (err instanceof Error) {
      throw customError({ id: 'Error', message: err.message });
    }
    return { id: 'Error', message: 'Internal error - errorParser() function.' } as ErrorType;
  }
};

/**
 * Logs a message to the console for debugging purposes, based on a specified status.
 * @param {string} log - The message to log.
 * @param {boolean} [status=false] - The status indicating whether to log the message.
 */
export const customDebug = (log: string, status: boolean = false) => {
  status && console.debug(log);
};
