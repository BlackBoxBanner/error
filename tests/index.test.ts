// Import the functions to be tested
import { customError, errorParser, customDebug } from '../src'; // Replace 'yourModule' with the actual module/file name

// Test customError function
describe('customError', () => {
  it('should create a custom error with default id', () => {
    const error = customError({ message: 'Test error message' });
    expect(error).toBeInstanceOf(Error);
    expect(error.message).toEqual(JSON.stringify({ id: 'internal', message: 'Test error message' }));
  });

  it('should create a custom error with specified id', () => {
    const error = customError({ id: 'customId', message: 'Test error message' });
    expect(error).toBeInstanceOf(Error);
    expect(error.message).toEqual(JSON.stringify({ id: 'customId', message: 'Test error message' }));
  });
});

// Test errorParser function
describe('errorParser', () => {
  it('should parse a valid custom error', () => {
    const inputError = new Error(JSON.stringify({ id: 'customId', message: 'Test error message' }));
    const parsedError = errorParser(inputError);
    expect(parsedError).toEqual({ id: 'customId', message: 'Test error message' });
  });

  it('should throw a custom error for invalid input', () => {
    const invalidInputError = new Error('Invalid JSON');
    expect(() => errorParser(invalidInputError)).toThrowErrorMatchingSnapshot();
  });
});

// Test customDebug function
describe('customDebug', () => {
  // Mock console.debug for testing
  const consoleDebugMock = jest.spyOn(console, 'debug').mockImplementation(() => {});

  // Clear the mock calls before each test
  beforeEach(() => {
    consoleDebugMock.mockClear();
  });

  it('should log the message when status is true', () => {
    customDebug('Test debug message', true);
    expect(consoleDebugMock).toHaveBeenCalledWith('Test debug message');
  });

  it('should not log the message when status is false', () => {
    customDebug('Test debug message', false);
    expect(consoleDebugMock).not.toHaveBeenCalled();
  });

  // Restore the original console.debug function after testing
  afterAll(() => {
    consoleDebugMock.mockRestore();
  });
});

