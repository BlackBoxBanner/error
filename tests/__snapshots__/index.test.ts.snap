// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`errorParser should throw a custom error for invalid input 1`] = `"{"id":"Error","message":"Unexpected token I in JSON at position 0"}"`;
